src/Facets/StargateFacetV2.sol

```diff
@@ -0,0 +1,166 @@
+
+    /// @notice Bridges tokens via Stargate Bridge
+    /// @param _bridgeData Data used purely for tracking and analytics
+    /// @param _stargateData Data specific to Stargate Bridge
+    function startBridgeTokensViaStargate(
+        ILiFi.BridgeData calldata _bridgeData,
+        StargateData calldata _stargateData
+    )
+        external
+        payable
+        nonReentrant
+        refundExcessNative(payable(msg.sender))
+        doesNotContainSourceSwaps(_bridgeData)
+        validateBridgeData(_bridgeData)
+    {
+        LibAsset.depositAsset(
+            _bridgeData.sendingAssetId,
+            _bridgeData.minAmount
+        );
+        _startBridge(_bridgeData, _stargateData);
+    }
+
+    /// @notice Performs a swap before bridging via Stargate Bridge
+    /// @param _bridgeData Data used purely for tracking and analytics
+    /// @param _swapData An array of swap related data for performing swaps before bridging
+    /// @param _stargateData Data specific to Stargate Bridge
+    function swapAndStartBridgeTokensViaStargate(
+        ILiFi.BridgeData memory _bridgeData,
+        LibSwap.SwapData[] calldata _swapData,
+        StargateData calldata _stargateData
+    )
+        external
+        payable
+        nonReentrant
+        refundExcessNative(payable(msg.sender))
+        containsSourceSwaps(_bridgeData)
+        validateBridgeData(_bridgeData)
+    {
+        _bridgeData.minAmount = _depositAndSwap(
+            _bridgeData.transactionId,
+            _bridgeData.minAmount,
+            _swapData,
+            payable(msg.sender),
+            _stargateData.fee.nativeFee
+        );
+
+        _startBridge(_bridgeData, _stargateData);
+    }
+
+    /// PRIVATE METHODS ///
+
+    /// @dev Contains the business logic for the bridging via StargateV2
+    /// @param _bridgeData Data used purely for tracking and analytics
+    /// @param _stargateData Data specific to Stargate Bridge
+    function _startBridge(
+        ILiFi.BridgeData memory _bridgeData,
+        StargateData calldata _stargateData
+    ) private {
+        _validateDestinationCallFlag(_bridgeData, _stargateData);
+
+        // get the address of the Stargate pool/router contract
+        address routerAddress = _getRouterAddress(_stargateData.assetId);
+
+        // check if NATIVE or ERC20
+        uint256 msgValue = _stargateData.fee.nativeFee;
+        if (LibAsset.isNativeAsset(_bridgeData.sendingAssetId)) {
+            // NATIVE
+            // add minAmount to msgValue
+            msgValue += _bridgeData.minAmount;
+        } else {
+            // ERC20
+            // check current allowance to router
+            ERC20 sendingAsset = ERC20(_bridgeData.sendingAssetId);
+            uint256 currentAllowance = sendingAsset.allowance(
+                address(this),
+                routerAddress
+            );
+            // check if allowance is sufficient
+            if (currentAllowance < _bridgeData.minAmount) {
```

Comment:
Why not just use `LibAsset::maxApproveERC20()` here? It does all of these checks anyway...

Reply:
Cause this is a gas-optimized version of achieving a max approval.
Main differences: 

- no check if sendingAssetId is != address(0) (as LibUtil does)
- no check if spender is != address(0) (as LibUtil does)
- middle step to approve to 0 only if current allowance is non-0 (LibUtil will do it in every case)